File: backend/src/main/java/com/finanziapp/backend/controller/TransaccionController.java
package com.finanziapp.backend.controller;

import java.util.List;

import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.CrossOrigin;
import org.springframework.web.bind.annotation.DeleteMapping;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.PutMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;

import com.finanziapp.backend.dto.BalanceResponse;
import com.finanziapp.backend.dto.TransaccionRequest;
import com.finanziapp.backend.entity.Transaccion;
import com.finanziapp.backend.service.TransaccionService;

import jakarta.validation.Valid;
import lombok.RequiredArgsConstructor;

@RestController
@RequestMapping("/api/transacciones")
@RequiredArgsConstructor
@CrossOrigin(origins = "*", allowedHeaders = "*")
public class TransaccionController {

    private final TransaccionService transaccionService;

    @PostMapping
    public ResponseEntity<?> crearTransaccion(@Valid @RequestBody TransaccionRequest request) {
        try {
            Transaccion guardada = transaccionService.registrarTransaccion(request);
            return ResponseEntity.status(HttpStatus.CREATED).body(guardada);
        } catch (IllegalArgumentException e) {
            return ResponseEntity.badRequest().body(e.getMessage());
        }
    }

    @GetMapping("/usuario/{idUsuario}")
    public ResponseEntity<List<Transaccion>> obtenerTransacciones(@PathVariable Long idUsuario) {
        return ResponseEntity.ok(transaccionService.obtenerPorUsuario(idUsuario));
    }

    @GetMapping("/balance/usuario/{idUsuario}")
    public ResponseEntity<BalanceResponse> obtenerBalance(@PathVariable Long idUsuario) {
        return ResponseEntity.ok(transaccionService.obtenerBalance(idUsuario));
    }

    @PutMapping("/{id}")
    public ResponseEntity<?> actualizarTransaccion(@PathVariable Long id, @Valid @RequestBody TransaccionRequest request) {
        try {
            Transaccion actualizada = transaccionService.actualizarTransaccion(id, request);
            return ResponseEntity.ok(actualizada);
        } catch (IllegalArgumentException e) {
            return ResponseEntity.badRequest().body(e.getMessage());
        }
    }

    @DeleteMapping("/{id}")
    public ResponseEntity<?> eliminarTransaccion(@PathVariable Long id) {
        try {
            transaccionService.eliminarTransaccion(id);
            return ResponseEntity.noContent().build();
        } catch (IllegalArgumentException e) {
            return ResponseEntity.badRequest().body(e.getMessage());
        }
    }
}

File: backend/src/modules/transacciones/transaccion.controller.ts
import { Request, Response, NextFunction } from 'express';
import { transaccionService } from './transaccion.service';
import { AppError } from '../../middlewares/errorHandler';

export class TransaccionController {
  async crearTransaccion(req: Request, res: Response, next: NextFunction): Promise<void> {
    try {
      const transaccion = await transaccionService.registrarTransaccion(req.body);
      res.status(201).json(transaccion);
    } catch (error) {
      next(error);
    }
  }

  async obtenerTransacciones(req: Request, res: Response, next: NextFunction): Promise<void> {
    try {
      const rawId = Array.isArray(req.params.idUsuario) ? req.params.idUsuario[0] : req.params.idUsuario;
      const idUsuario = parseInt(rawId, 10);
      if (isNaN(idUsuario)) {
        throw new AppError('El ID de usuario debe ser un número válido', 400);
      }

      const transacciones = await transaccionService.obtenerPorUsuario(idUsuario);
      res.status(200).json(transacciones);
    } catch (error) {
      next(error);
    }
  }

  async obtenerBalance(req: Request, res: Response, next: NextFunction): Promise<void> {
    try {
      const rawId = Array.isArray(req.params.idUsuario) ? req.params.idUsuario[0] : req.params.idUsuario;
      const idUsuario = parseInt(rawId, 10);
      if (isNaN(idUsuario)) {
        throw new AppError('El ID de usuario debe ser un número válido', 400);
      }

      const balance = await transaccionService.obtenerBalance(idUsuario);
      res.status(200).json(balance);
    } catch (error) {
      next(error);
    }
  }

  async actualizarTransaccion(req: Request, res: Response, next: NextFunction): Promise<void> {
    try {
      const rawId = Array.isArray(req.params.id) ? req.params.id[0] : req.params.id;
      const id = parseInt(rawId, 10);
      if (isNaN(id)) {
        throw new AppError('El ID de transacción debe ser un número válido', 400);
      }

      const actualizada = await transaccionService.actualizarTransaccion(id, req.body);
      res.status(200).json(actualizada);
    } catch (error) {
      next(error);
    }
  }

  async eliminarTransaccion(req: Request, res: Response, next: NextFunction): Promise<void> {
    try {
      const rawId = Array.isArray(req.params.id) ? req.params.id[0] : req.params.id;
      const id = parseInt(rawId, 10);
      if (isNaN(id)) {
        throw new AppError('El ID de transacción debe ser un número válido', 400);
      }

      await transaccionService.eliminarTransaccion(id);
      res.status(204).send();
    } catch (error) {
      next(error);
    }
  }
}

export const transaccionController = new TransaccionController();
